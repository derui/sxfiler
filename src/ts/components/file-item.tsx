import classNames from "classnames";
import * as React from "react";
import * as Domain from "../domain/node";
import FileMode from "./file-item-mode";
import FileName from "./file-item-name";
import FileSize from "./file-item-size";
import FileTimestamp from "./file-item-timestamp";

interface Prop {
  item: Domain.Node;
  marked: boolean;
  selected: boolean;
}

export default class FileItem extends React.PureComponent<Prop> {
  public render() {
    const { item } = this.props;
    const className = classNames("fp-FileItem", {
      "fp-FileItem-selected": this.props.selected,
      "fp-FileItem-marked": this.props.marked,
    });

    return (
      <li className={className}>
        <FileMode
          key="mode"
          mode={item.stat.mode}
          isDirectory={item.stat.isDirectory}
          isSymlink={item.stat.isSymlink}
        />
        <FileTimestamp key="timestamp" timestamp={item.stat.mtime} />
        <FileSize key="size" size={item.stat.size} />
        <FileName key="name" name={item.name} isDirectory={item.stat.isDirectory} isSymlink={item.stat.isSymlink} />
      </li>
    );
  }
}
