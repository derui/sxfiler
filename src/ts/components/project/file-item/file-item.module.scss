@import "../../ui/commom-style/_colors.scss";
@import "../../ui/commom-style/_variables.scss";
@import "../../ui/commom-style/reset.module.css";

.fileItem {
  composes: listItem-base from "../../ui/list-item/list-item.module.scss";

  padding: $baseSpace 0;
  width: 100%;

  background-color: $sol_base03;

  font-family: monospace;
  color: $sol_base1;
  white-space: nowrap;

  font-size: 1rem;
  border-left: 4px solid transparent;

  &[aria-selected="true"] {
    border-left: 4px solid $sol_orange;
  }

  &[aria-hidder="true"] {
    visibility: hidden;
  }

  &[data-marked="true"] {
    background-color: rgba($sol_blue, 0.25);
  }
}


.timestamp {
  flex: 0 1 auto;
  padding: 0 $smallSpace;
  text-align: right;

  white-space: nowrap;
}

.name {
  flex: 1 0 auto;
  padding: 0;
  font-weight: bold;
}

.name.nameDirectory {
  color: $sol_yellow;
}

.name.nameSymlink {
  color: $sol_orange;
}

.size {
  flex: 0 1 auto;
  padding: 0 $baseSpace;
  text-align: right;

  white-space: pre;
}

.mode {
  flex: 0 1 auto;
  padding: 0 $smallSpace;

  white-space: nowrap;
}
