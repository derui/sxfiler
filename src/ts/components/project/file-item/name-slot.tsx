import classNames from "classnames";
import * as React from "react";

// eslint-disable-next-line @typescript-eslint/no-var-requires
const styles = require("./file-item.module.scss");

interface Prop {
  isDirectory: boolean;
  isSymlink: boolean;
  name: string;
}

const NodeName: React.FC<Prop> = ({ isDirectory, name, isSymlink }) => {
  const className = classNames(styles.name, {
    [styles.nameDirectory]: isDirectory && !isSymlink,
    [styles.nameSymlink]: isSymlink,
  });

  return <span className={className}>{name}</span>;
};

export default NodeName;
