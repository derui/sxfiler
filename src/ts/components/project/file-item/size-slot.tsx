import bigInt from "big-integer";
import * as React from "react";

// eslint-disable-next-line @typescript-eslint/no-var-requires
const styles = require("./file-item.module.scss");

enum SizeUnit {
  Byte = "byte",
  KByte = "kbyte",
  MByte = "mbyte",
  GByte = "gbyte",
  TByte = "tbyte",
  Unknown = "unknown",
}

interface SizeData {
  sizeUnit: SizeUnit;
  size: bigInt.BigInteger;
  alignedSize: number;
}

function nextUnit(sizeUnit: SizeUnit): SizeUnit | null {
  switch (sizeUnit) {
    case "byte":
      return SizeUnit.KByte;
    case "kbyte":
      return SizeUnit.MByte;
    case "mbyte":
      return SizeUnit.GByte;
    case "gbyte":
      return SizeUnit.TByte;
    case "tbyte":
      return SizeUnit.Unknown;
    default:
      return null;
  }
}

function sizeUnitToString(sizeUnit: SizeUnit): string {
  switch (sizeUnit) {
    case "byte":
      return "B";
    case "kbyte":
      return "K";
    case "mbyte":
      return "M";
    case "gbyte":
      return "G";
    case "tbyte":
      return "T";
    default:
      return "-";
  }
}

class Size {
  private data: SizeData;

  constructor(size: bigInt.BigInteger) {
    this.data = this.toData(size);
  }
  /**
   * convert to String
   */
  public toString(): string {
    const unit = sizeUnitToString(this.data.sizeUnit);
    const fixed = this.data.alignedSize.toFixed(1);
    const padded = fixed.padStart(5, " ");
    return `${padded}${unit}`;
  }

  /**
   * parse size string to data structure
   * @param size string representation of size
   * @return size data
   */
  private toData(size: bigInt.BigInteger): SizeData {
    function calcUnit(
      fileSize: bigInt.BigInteger,
      current: SizeUnit,
      decimal: number
    ): { current: SizeUnit; size: number } {
      if (bigInt.zero.leq(fileSize) && fileSize.lt(bigInt(1024))) {
        return { current, size: fileSize.toJSNumber() };
      }

      const unit = nextUnit(current);
      if (unit == null) {
        return { current, size: fileSize.toJSNumber() + decimal };
      } else {
        return calcUnit(fileSize.divide(bigInt(1024)), unit, fileSize.mod(bigInt(1024)).toJSNumber() / 1024);
      }
    }
    const parsedSize = bigInt(size);
    const data = calcUnit(parsedSize, SizeUnit.Byte, 0);

    return {
      sizeUnit: data.current,
      alignedSize: data.size,
      size: parsedSize,
    };
  }
}

interface Prop {
  size: bigInt.BigInteger;
}

const NodeSize: React.FC<Prop> = prop => {
  const size = new Size(prop.size);

  return <pre className={styles.size}>{size.toString()}</pre>;
};

export default NodeSize;
