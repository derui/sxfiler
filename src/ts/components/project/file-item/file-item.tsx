import * as React from "react";
import * as Domain from "../../../domains/file-item";
import * as ListItem from "../../ui/list-item/list-item";

import Mode from "./mode-slot";
import Name from "./name-slot";
import Size from "./size-slot";
import Timestamp from "./timestamp-slot";
import { ForwardedRef } from "../../ui/util";

// eslint-disable-next-line @typescript-eslint/no-var-requires
const styles = require("./file-item.module.scss");

export type Props = ForwardedRef & {
  item: Domain.FileItem;
  selected: boolean;
  hidden?: boolean;
};

const Element = ListItem.createComponent();

export class Component extends React.PureComponent<Props> {
  public render() {
    const { item, selected, hidden = false, ...rest } = this.props;

    return (
      <Element
        className={styles.fileItem}
        aria-selected={selected}
        data-marked={item.marked}
        aria-hidden={hidden}
        {...rest}
      >
        <Mode key="mode" mode={item.stat.mode} isDirectory={item.stat.isDirectory} isSymlink={item.stat.isSymlink} />
        <Timestamp key="timestamp" timestamp={item.stat.mtime} />
        <Size key="size" size={item.stat.sizeAsBigInt()} />
        <Name key="name" name={item.name} isDirectory={item.stat.isDirectory} isSymlink={item.stat.isSymlink} />
      </Element>
    );
  }
}
