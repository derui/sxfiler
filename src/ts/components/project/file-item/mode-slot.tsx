import * as React from "react";
import { styled } from "@/components/theme";
import { CapabilityObject } from "@/domains/capability";
import { ModeObject } from "@/domains/mode";

function capabilityToString(cap: CapabilityObject) {
  const readable = cap.readable ? "r" : "-";
  const writable = cap.writable ? "w" : "-";
  const executable = cap.executable ? "x" : "-";

  return `${readable}${writable}${executable}`;
}

/**
 * convert mode to string
 */
function modeToString(mode: ModeObject, isDirectory: boolean, isSymlink: boolean): string {
  let state = "-";

  if (isDirectory && !isSymlink) {
    state = "d";
  } else if (isSymlink) {
    state = "l";
  }

  const owner = capabilityToString(mode.owner);
  const group = capabilityToString(mode.group);
  const others = capabilityToString(mode.others);

  return `${state}${owner}${group}${others}`;
}

interface Prop {
  mode: ModeObject;
  isDirectory: boolean;
  isSymlink: boolean;
}

const Mode = styled.span`
  flex: 0 1 auto;
  padding: 0 ${props => props.theme.spaces.small};

  white-space: nowrap;
`;

export const Component: React.FC<Prop> = prop => {
  const data = modeToString(prop.mode, prop.isDirectory, prop.isSymlink);

  return <Mode>{data}</Mode>;
};
