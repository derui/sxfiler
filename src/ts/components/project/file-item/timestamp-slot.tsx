import * as React from "react";
import { styled } from "@/components/theme";

export type Props = {
  timestamp: Date;
};

/**
 * format Date to display in item
 */
const format = function format(timestamp: Date): string {
  const year = `${timestamp.getFullYear()}`.padStart(4, "0");
  const month = `${timestamp.getMonth() + 1}`.padStart(2, "0");
  const date = `${timestamp.getDate()}`.padStart(2, "0");

  return `${year}/${month}/${date}`;
};

const Timestamp = styled.span`
  flex: 0 1 auto;
  padding: 0 ${props => props.theme.spaces.small};
  text-align: right;

  white-space: nowrap;
`;

export const Component: React.FC<Props> = prop => {
  const date = format(prop.timestamp);
  return <Timestamp>{date}</Timestamp>;
};
