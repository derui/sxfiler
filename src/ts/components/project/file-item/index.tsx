import { h } from "preact";

import { Component as Mode } from "./mode-slot";
import { Component as Name } from "./name-slot";
import { Component as Size } from "./size-slot";
import { Component as Timestamp } from "./timestamp-slot";
import { FileItem } from "@/generated/filer_pb";

export type Props = {
  item: FileItem.AsObject;
  selected: boolean;
  hidden?: boolean;
  bookmarked: boolean;
  onRefUpdated?: (e: HTMLElement) => void;
};

export const Component: preact.FunctionComponent<Props> = ({
  item,
  selected,
  bookmarked,
  hidden = false,
  onRefUpdated = undefined,
  ...rest
}) => {
  const stat = item.stat;
  if (!stat) {
    return null;
  }

  return (
    <div
      class="file-item__root"
      data-testid="fileList-item"
      aria-selected={selected}
      data-marked={item.marked}
      data-bookmarked={bookmarked}
      aria-hidden={hidden}
      ref={(e) => {
        if (onRefUpdated && e) {
          onRefUpdated(e);
        }
      }}
      {...rest}
    >
      <Mode key="mode" mode={stat.mode} isDirectory={stat.isDirectory} isSymlink={stat.isSymlink} />
      <Timestamp key="timestamp" timestamp={new Date(stat.mtime)} />
      <Size key="size" size={BigInt(stat.size)} />
      <Name key="name" name={item.name} isDirectory={stat.isDirectory} isSymlink={stat.isSymlink} />
    </div>
  );
};
Component.displayName = "FileItem";

export type ComponentType = typeof Component;
