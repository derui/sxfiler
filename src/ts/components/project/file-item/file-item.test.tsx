import * as React from "react";
import renderer from "react-test-renderer";

import { createFileItem } from "../../../domains/file-item";

import { Component as T } from "./file-item";
import { createMode } from "../../../domains/mode";
import { emptyCapability } from "../../../domains/capability";
import { createFileStat } from "../../../domains/file-stat";

function makeNode(marked: boolean, isDirectory = false, isSymlink = false) {
  return createFileItem({
    id: "node",
    name: "file.ext",
    marked,
    stat: createFileStat({
      mode: createMode({
        owner: emptyCapability()
          .allowToRead()
          .allowToWrite(),
        group: emptyCapability().allowToRead(),
        others: emptyCapability().allowToRead(),
      }),
      uid: 1000,
      gid: 1000,
      atime: "0",
      ctime: "0",
      mtime: "0",
      size: "10",
      isDirectory: isDirectory,
      isFile: !isDirectory && !isSymlink,
      isSymlink: isSymlink,
    }),
    parentDirectory: "/",
  });
}

describe("Project", () => {
  describe("Node Item", () => {
    it("should print correctly", () => {
      const node = makeNode(false);
      const tree = renderer.create(<T item={node} selected={false} />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it("should add class correctly when marked", () => {
      const node = makeNode(true);
      const tree = renderer.create(<T item={node} selected={false} />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it("should add class correctly when selected", () => {
      const node = makeNode(false);
      const tree = renderer.create(<T item={node} selected={true} />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it("should add class correctly when selected and marked", () => {
      const node = makeNode(true);
      const tree = renderer.create(<T item={node} selected={true} />).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });
});
