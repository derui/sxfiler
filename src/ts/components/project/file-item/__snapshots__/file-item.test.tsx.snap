// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project Node Item should add class correctly when marked 1`] = `
<div
  aria-hidden={false}
  aria-selected={false}
  className="nodeItem"
  data-marked={true}
  role="listitem"
>
  <span
    className="mode"
  >
    -rw-r--r--
  </span>
  <span
    className="timestamp"
  >
    1970/01/01
  </span>
  <pre
    className="size"
  >
     10.0B
  </pre>
  <span
    className="name"
  >
    file.ext
  </span>
</div>
`;

exports[`Project Node Item should add class correctly when selected 1`] = `
<div
  aria-hidden={false}
  aria-selected={true}
  className="nodeItem"
  data-marked={false}
  role="listitem"
>
  <span
    className="mode"
  >
    -rw-r--r--
  </span>
  <span
    className="timestamp"
  >
    1970/01/01
  </span>
  <pre
    className="size"
  >
     10.0B
  </pre>
  <span
    className="name"
  >
    file.ext
  </span>
</div>
`;

exports[`Project Node Item should add class correctly when selected and marked 1`] = `
<div
  aria-hidden={false}
  aria-selected={true}
  className="nodeItem"
  data-marked={true}
  role="listitem"
>
  <span
    className="mode"
  >
    -rw-r--r--
  </span>
  <span
    className="timestamp"
  >
    1970/01/01
  </span>
  <pre
    className="size"
  >
     10.0B
  </pre>
  <span
    className="name"
  >
    file.ext
  </span>
</div>
`;

exports[`Project Node Item should print correctly 1`] = `
<div
  aria-hidden={false}
  aria-selected={false}
  className="nodeItem"
  data-marked={false}
  role="listitem"
>
  <span
    className="mode"
  >
    -rw-r--r--
  </span>
  <span
    className="timestamp"
  >
    1970/01/01
  </span>
  <pre
    className="size"
  >
     10.0B
  </pre>
  <span
    className="name"
  >
    file.ext
  </span>
</div>
`;
